Fix and standardize multiple table UI inconsistencies

alexkim205 opened the pull request:
## Changes

Wrapped up a bunch of table fixes in this PR.

- Fixes #4725 (Master) Table Overflow Revamp
- Fixes #4500 Feature flags table formatted wrong on the first load
- Fixes #4721 Squishy tables
- Fixes #4723 Actions Table Overflow inaccessible on smaller screens
- Fixes #4724 Inconsistent scroll shading for fully visible table

## A bunch of after pics (in no particular order)

<img width="1068" alt="Screen Shot 2021-06-11 at 4 39 55 PM" src="https://user-images.githubusercontent.com/13460330/121764462-f8cfcc00-caf8-11eb-9310-8d79f2fae3bb.png">

https://user-images.githubusercontent.com/13460330/121764458-efdefa80-caf8-11eb-90c5-c7304560a192.mov

https://user-images.githubusercontent.com/13460330/121764445-c6be6a00-caf8-11eb-87c9-130563d65905.mov

<img width="1160" alt="Screen Shot 2021-06-11 at 8 55 22 PM" src="https://user-images.githubusercontent.com/13460330/121764451-d9d13a00-caf8-11eb-8700-e314d105c68f.png">


## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [X] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] Frontend/CSS is usable at 320px (iPhone SE) and decent at 360px (most phones)

alexkim205 committed:
fix multiple table bugs and standardize table overflows

mariusandra commented on frontend/src/lib/utils/responsiveUtils.ts:
```diff
@@ -1,5 +1,14 @@
+        [key]: parsePixelValue(cssStatement),
+    }),
+    {}
+)
```
Let's avoid using `reduce` whenever possible. The only thing using `reduce` shows to whoever is reading the code is that the author is proficient in using `reduce`, _nothing_ else :). It's usually longer and always less readable (`==` maintainable) than all other ways of doing the same thing.

The following does the same, but is much simpler to understand:

```js
const BREAKPOINT_MAP = Object.fromEntries(
    Object.entries(responsiveMap).map(([key, cssStatement]) => [key, parsePixelValue(cssStatement)])
)
```

... and is most likely much faster as well (no need to create and spread `array.length` temporary objects).

alexkim205 replied:
super clean, thanks for the tip :)

mariusandra commented on frontend/src/lib/utils/responsiveUtils.ts:
```diff
@@ -16,12 +25,15 @@ export function parsePixelValue(cssStatement: string): number {
         }
     })
     return breakpoint
 }
```
I'd also simplify this to:

```js
export function getActiveBreakpoint(): number {
    const width = window.innerWidth
    const lastMatchingBreakpoint = BREAKPOINT_VALUES.filter(value => width >= value).pop()
    return lastMatchingBreakpoint || BREAKPOINT_VALUES[0]
}
```

mariusandra replied:
I also renamed this function `getActiveBreakpointValue` (over `getActiveBreakpoint`) since technically a "breakpoint" is one of "sm", "xs", etc, not the value in pixels it resolves to.

And I changed `width > value` to `width >= value`, as that's how (antd) breakpoints [are defined](https://ant.design/components/grid/#Col).

mariusandra committed:
support more es2019 functions like Object.fromEntries

mariusandra committed:
simplify getting active breakout value

mariusandra reviewed (changes requested):
I simplified the breakpoint code a bit, yet still found several issues in the app that I'll gladly pass on :)

1. Feature flags do not display well, no matter how I open the page. The key is always cut for me, and the description area is needlessly huge:
![image](https://user-images.githubusercontent.com/53387/121871887-ed50f080-cd04-11eb-9a29-be9f5d9b7037.png)

2. Events page (zoomed out to simulate a wide screen) won't let me resize things like I expect:
![2021-06-14 11 39 00](https://user-images.githubusercontent.com/53387/121872218-4e78c400-cd05-11eb-89df-5d4f3777e23c.gif)

3. No "you can scroll right" indicator here:
![2021-06-14 11 41 20](https://user-images.githubusercontent.com/53387/121872393-7f58f900-cd05-11eb-89cf-57fc13334e2d.gif)

Probably outside the scope, or maybe not:

4. If the widths auto-update while resizing, my dragging cursor is "let go" and I need to start resizing again:

![2021-06-14 11 44 43](https://user-images.githubusercontent.com/53387/121872829-06a66c80-cd06-11eb-9738-a23cceeb33f7.gif)

5. There's no way to make the last column of any table wider or shorter, plus the whitespace that's at the end of the table:
![2021-06-14 11 46 49](https://user-images.githubusercontent.com/53387/121873061-440afa00-cd06-11eb-820d-46fdb801d3da.gif)

---

PS, my most common PR review comment is "simplify!", so feel free to always do that automatically :D. The logic here is that code is *not* written for computers to read, but for other humans to read. Computers can read your code no matter how you well you write it, but the human reading it will be caught off guard by clever code that's hard to parse. So always optimise code clarity for the cognitively lowest life form in the process: your fellow humans. Usually this means "make your code so simple it looks like a first year student wrote it".

mariusandra commented:
The breaking cypress tests seem relevant as well.

mariusandra commented:
One more bug, this time in [this](https://app.posthog.com/insights?insight=TRENDS&interval=day&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%5D&filter_test_accounts=false&properties=%5B%5D&display=ActionsTable) cohorts table:
![2021-06-14 13 00 02](https://user-images.githubusercontent.com/53387/121882173-7d486780-cd10-11eb-8bb3-7052b86b8ad1.gif)

paolodamico commented:
Looks like @mariusandra already did a comprehensive review of this, tag me for a follow up review if needed @alexkim205

samwinslow committed:
unset maxWidth for resizable cols by default

samwinslow commented:
> 2. Events page (zoomed out to simulate a wide screen) won't let me resize things like I expect:

Fixed this by removing the arbitrary max column width I had set by default.

> 1. There's no way to make the last column of any table wider or shorter, plus the whitespace that's at the end of the table:

This one gets tricky because in order for the table to render correctly, one column needs to have an uncontrolled width set by the browser. But, because the size of the entire table and other columns update together, it works out such that this "uncontrolled" column appears to have a fixed width. Right now the last column is chosen for this unfortunate caveat, but maybe we could always push a blank column onto the end of the ones given to us via props.

mariusandra commented:
> In order for the table to render correctly, one column needs to have an uncontrolled width set by the browser.

Why? :)

samwinslow commented:
@mariusandra If every column has a set width, the browser will just decide what it thinks appropriate width values are in order to fill the horizontal space of the `<table>`, when using default `table-layout`... Antd's `tableLayout: fixed` prop did not seem to change this behavior but more experimentation could be done. 

Antd's docs have a similar warning:
> Specify width of columns if header and cell do not align properly. If specified width is not working or have gutter between columns, please try to leave one column at least without width to fit fluid layout, or make sure no long word to break table layout.

When I was working on ResizableTable it was very hard to keep the resizable column headers in sync with the table's actual columns. I don't mean to suggest it's impossible but it might get hacky!

mariusandra commented:
Yeah, this sounds tricky. I don't know the nitty gritty details of antd's tables, and those tradeoffs probably make sense.

Yet it's funny how "resizable table headers" is still an unsolved computer science problem :).

Can't we just, after the table has fully loaded, capture all the flexibly set column widths, and then just keep them all in a controlled state as we start to resize? Not sure if this helps anything 😅

samwinslow commented:
@alexkim205 The example in the Antd docs is a bit of a red herring — it re-renders the entire table on each resize event, which happens every few milliseconds. It gets borked on very long tables like Events or Sessions. Even if the resizes are throttled to occur once every 60 or 100 ms, e.g. with `requestAnimationFrame`, rendering may still take longer, and a 100ms interval works out to about 10 frames per second — looks quite bad.

This actually was my first approach, but then I settled on a CSS-based solution; updating the widths specified in `<colgroup>` styles does not trigger a React render, but only a browser repaint.

alexkim205 commented:
Discussed some possible approaches during standup. Here are some notes for posterity:

- passing in `columns` prop with width makes antd's `<Table>` component rerender aggressively
- Virtual table header is a good solution but introduces several bugs like the one this PR tries to address
- Maybe see if we can use other table libraries that have better handling of large amounts of data
   - [X] `react-table` - [CodeSandbox](https://codesandbox.io/s/nice-christian-4gkhr) doesn't handle many rows well either.

mariusandra commented:
`react-table` looks pretty awesome! Unless I'm terribly wrong, it shouldn't be that hard to implement it and make our tables look like they do now? :)

alexkim205 commented:
Yup it's a headless utility library so it's pretty cool that you can pick and choose features you want to include. One feature I've been having a hard time playing nicely is expandability. Still trying to figure out a way to make it extensible while using `react-window`'s FixedSizeList. 

In summary, want to combine 
* https://react-table.tanstack.com/docs/examples/virtualized-rows
* https://react-table.tanstack.com/docs/examples/expanding
* https://react-table.tanstack.com/docs/examples/full-width-resizable-table <--- example is broken on their website

alexkim205 committed:
fix feature flag key cutoff and empty description bug

alexkim205 commented:
> 1. Feature flags do not display well, no matter how I open the page. The key is always cut for me, and the description area is...

1. Addressed bug 1 by ellipsifying key and description columns (the two most variable columns) for more consistent row heights. Added tooltips so that user can still view any text that is cutoff. 
2. Fixed failing cypress tests with more specific selectors
3. Fixed column widths of action columns (Enabled, Usage, Actions) so that we can have more space for the other columns and fixed them to the right.

### Demo


https://user-images.githubusercontent.com/13460330/122113089-7e43cc80-cdd6-11eb-80cf-f7ff3249c327.mov



Bugs 2-5 happen on the events table and I think it requires a different implementation of the events table to fix this. I'll address this in another PR as it's out of the scope.

alexkim205 committed:
ellipsify feature flag table

alexkim205 commented:
> One more bug, this time in [this](https://app.posthog.com/insights?insight=TRENDS&interval=day&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%5D&filter_test_accounts=false&properties=%5B%5D&display=ActionsTable) cohorts table:
> ![2021-06-14 13 00 02](https://user-images.githubusercontent.com/53387/121882173-7d486780-cd10-11eb-8bb3-7052b86b8ad1.gif)

I think this bug is fixed by my commit here that was merged in as part of @paolodamico's InsightsTableV2 update. But the fix hasn't been made in V1, so I'll bring my commit over here too.

alexkim205 committed:
labels or dates or days

alexkim205 commented:
Whelp, should have merged instead of rebase from master. Please ignore all those commits that I brought in except the last one. I'll squash and clean up accordingly when I merge this PR in.

mariusandra commented:
Uh... sorry, but I can't review this. Not sure how it came to be, but there are thousands of changed lines in 77 files:

![image](https://user-images.githubusercontent.com/53387/122202806-ac401400-ce9d-11eb-9a33-02a299a41b01.png)

Can you please clean it up? :)

mariusandra reviewed (changes requested):
.

alexkim205 commented:
Hey @mariusandra sorry about the commit cereal. I think this rev should be cleaner

mariusandra committed:
Merge branch 'master' into fix/some-table-bugs

mariusandra commented on frontend/src/lib/components/Table.tsx:
```diff
@@ -7,6 +7,7 @@ import { TZLabel } from './TimezoneAware'
 export function createdAtColumn(): Record<string, any> {
     return {
         title: 'Created',
+        ellipsis: true,
```
Instead of an ellipsis, can't we just "nowrap" this column? The "2 days ago" texts are usually very small, and just having a random "..." in the end feels weird:

![image](https://user-images.githubusercontent.com/53387/122371736-60a26e80-cf60-11eb-8e9c-c8796e2e1528.png)

alexkim205 replied:
agreed, good shout!

alexkim205 replied:
Something like this?

<img width="624" alt="Screen Shot 2021-06-21 at 9 44 18 AM" src="https://user-images.githubusercontent.com/13460330/122798294-42e14c00-d275-11eb-9a12-1d4572d01e04.png">

mariusandra replied:
Hmm, not really. I meant to just expand the column to fit everything... 🙃

alexkim205 replied:
Ah how does this look

<img width="513" alt="Screen Shot 2021-06-21 at 10 40 53 AM" src="https://user-images.githubusercontent.com/13460330/122804810-306b1080-d27d-11eb-9557-9278be872517.png">

mariusandra commented on frontend/src/lib/components/Table.tsx:
```diff
@@ -19,6 +20,7 @@ export function createdByColumn(items: Record<string, any>[]): Record<string, an
     const { user } = useValues(userLogic)
     return {
         title: 'Created by',
+        width: 150,
```
I'm also slightly vary about imposing this. You might have a long (e.g. indian, russian, whatever) name that just breaks with "..." on the last letter. 

I'd be a much bigger fan of something like: "max-width: 300px, width: auto", though that might be annoyingly complicated to implement?

alexkim205 replied:
Got it let's do that! It doesn't look too complicated actually

mariusandra reviewed (dismissed):
I'm not sure how annoying this feedback is to fix, but it's all given in the service of "let's have better UX". I'm also not sure if any of my comments override anything said before or not. :)

Some comments inline, some below.

1. Let's remove the ellipsis from the feature flags table. I think There's no reason this table needs fixed heights. Looking at [our own data](https://app.posthog.com/feature_flags), if we'd merge this, we'd have a sub-par experience... as we'll have mountains of whitespace all around the page, yet we won't be able to actually read any flag keys and descriptions. That sounds like a failure. So can we just make it look like this, but... uhm... work? :D

![image](https://user-images.githubusercontent.com/53387/122370207-2be1e780-cf5f-11eb-9ee3-bdf9a62294b9.png)

2. Feature flags don't tick the "320px" checkbox: 
![2021-06-17 11 23 52](https://user-images.githubusercontent.com/53387/122369514-8d558680-cf5e-11eb-985f-a88ab95da6d2.gif)

3. Similar comment for the annotations table. The annotation column is anyway in two lines sometimes if "last updated" pushes it so.... so why the ellipsis? Let's remove it. The whitespace is also very suboptimal in full screen.

![image](https://user-images.githubusercontent.com/53387/122370857-b4608800-cf5f-11eb-806d-1ded0c823178.png)
![image](https://user-images.githubusercontent.com/53387/122370934-c2160d80-cf5f-11eb-8eb3-1db5d3acbc48.png)

alexkim205 commented:
Addressed a-c above and added some goodies where the table doesn't word break columns when entering scroll mode. 

**Annotations Table**

https://user-images.githubusercontent.com/13460330/122806700-7de87d00-d27f-11eb-894f-6a933454af55.mov

**Feature Flags Table**

https://user-images.githubusercontent.com/13460330/122806731-8a6cd580-d27f-11eb-8860-6e972edea0b1.mov

**Word breaking for date-specific columns**
<img width="513" alt="Screen Shot 2021-06-21 at 10 40 53 AM" src="https://user-images.githubusercontent.com/13460330/122806875-bc7e3780-d27f-11eb-8228-6c3ea6ec5e94.png">


Failing e2e tests don't seem related

alexkim205 committed:
fix bugs; refactor table scroll

alexkim205 committed:
Merge remote-tracking branch 'origin' into fix/some-table-bugs

alexkim205 committed:
Merge branch 'master' into fix/some-table-bugs

mariusandra committed:
Merge branch 'master' into fix/some-table-bugs

mariusandra committed:
fix double key and increase the width just a bit

mariusandra commented:
Found (and fixed) just one weirdness with a double key for feature flags, probably from some merge:
![image](https://user-images.githubusercontent.com/53387/123281969-683ab800-d50a-11eb-8f12-05b8d291cb70.png)

Other than that, there's one new weirdness with the events table. Yet I'd suggest to ~~already~~finally merge this and tackle this in a new issue.

![2021-06-24 16 46 20](https://user-images.githubusercontent.com/53387/123283628-ddf35380-d50b-11eb-8595-806b1e959c38.gif)

mariusandra commented:
Even got it to look like this one time:
![image](https://user-images.githubusercontent.com/53387/123285214-24957d80-d50d-11eb-8c09-f442206d21b3.png)

alexkim205 commented:
Thanks for the hot fix, not sure why that happened. Yeah the events table (and other tables that use our custom ResizableTable component) is a bit funky and has a few bugs that I think should be addressed in another PR. `react-table` looks like the best approach so far
